app/domain/authentication/authn_azure/validate_application_identity.rb

```diff
@@ -0,0 +1,102 @@
+require 'command_class'
+
+module Authentication
+  module AuthnAzure
+
+    Log = LogMessages::Authentication
+    Err = Errors::Authentication::AuthnAzure
+    # Possible Errors Raised: RoleNotFound, InvalidApplicationIdentity
+
+    ValidateApplicationIdentity = CommandClass.new(
+        dependencies: {
+            role_class:                 Role,
+            resource_class:             Resource,
+            application_identity_class: ApplicationIdentity,
+            logger:                     Rails.logger
+        },
+        inputs: %i(account service_id username xms_mirid oid)
+    ) do
+
+      def call
+        parse_xms_mirid
+        # compare xms_mirid with what is defined in annotations
+        validate_application_identity
+      end
+
+      private
+
+      # validate integrity of annotations
+      def application_identity
+        @application_identity ||= @application_identity_class.new(
+            role_annotations: role_annotations,
+            service_id: @service_id
+        )
+      end
+
+      def role_annotations
+        role.annotations
+      end
+
+      # convert host ID to full host ID, cucumber:host:somepolicy
+      # TODO For a Conjur user, an incorrect full id will be returned
+      def role_id
+        @role_id ||= @role_class.roleid_from_username(@account, @username)
+      end
+
+      def role
+        @role ||= @resource_class[role_id]
+        raise SecurityErr::RoleNotFound(role_id) if @username.nil?
+        return @role
+      end
+
+      # TODO add parsing format to check xms mirid is in proper format
+      # xms_mirid is a term in Azure to define a claim that describes the resource that holds the encoding of the instance's
+      # subscription, resource group, and provider identity needed for authorization. xms_mirid is one of the fields in
+      # the JWT token. This function will extract the relevant information from xms_mirid claim and populate a representative
+      # hash with the appropriate fields.
+      def parse_xms_mirid
```

Comment:
we usually call methods in the name of the returned object. I think we should have a method named `token_identity_from_xms_mirid` (or something) which will use memoization to create an object.

Reply:
what about `token_identity_from_claims` b/c it is not only `xms_mirid` that is being used to build the `token_identity` object (also `oid` if system-assigned-identity). This might play into Jonah's comment about creating two separate objects...

Reply:
cool.